internal/admin/handle_health_authority.go

```diff
@@ -0,0 +1,240 @@
+				return
+			}
+		}
+		m["ha"] = healthAuthority
+		m["hak"] = &model.HealthAuthorityKey{From: time.Now()} // For create form.
+		c.HTML(http.StatusOK, "healthauthority", m)
+	}
+}
+
+// HandleHealthAuthorityKeys handles the keys action for health authorities.
+func (s *Server) HandleHealthAuthorityKeys() func(c *gin.Context) {
+	return func(c *gin.Context) {
+		ctx := c.Request.Context()
+
+		haDB := database.New(s.env.Database())
+		haID, err := strconv.ParseInt(c.Param("id"), 10, 64)
+		if err != nil {
+			ErrorPage(c, "Unable to parse `id` param")
+			return
+		}
+		healthAuthority, err := haDB.GetHealthAuthorityByID(ctx, haID)
+		if err != nil {
+			ErrorPage(c, fmt.Sprintf("error processing health authority: %v", err))
+			return
+		}
+
+		if action := c.Param("action"); action == "create" {
+			var form keyhealthAuthorityFormData
+			err := c.Bind(&form)
+			if err != nil {
+				ErrorPage(c, err.Error())
+				return
+			}
+
+			var hak model.HealthAuthorityKey
+			err = form.PopulateHealthAuthorityKey(&hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error parsing new health authority key: %v", err))
+				return
+			}
+			err = haDB.AddHealthAuthorityKey(ctx, healthAuthority, &hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error saving health authority key: %v", err))
+				return
+			}
+
+		} else if action == "revoke" || action == "reinstate" || action == "activate" {
+			version := c.Param("version")
+
+			// find the key.
+			var hak *model.HealthAuthorityKey
+			for _, key := range healthAuthority.Keys {
+				if key.Version == version {
+					hak = key
+					break
+				}
+			}
+			if hak == nil {
+				ErrorPage(c, "Invalid key specified")
+				return
+			}
+
+			if action == "activate" {
+				if hak.IsFuture() {
+					hak.From = time.Now()
+				}
+			} else if action == "revoke" {
+				hak.Revoke()
+			} else {
+				hak.Thru = time.Time{}
+			}
+
+			log.Printf("HAK %+v", hak)
+
+			err = haDB.UpdateHealthAuthorityKey(ctx, hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error saving health authority key: %v", err))
+				return
+			}
+
+		} else {
+			ErrorPage(c, "invalid action")
+			return
+		}
+		c.Redirect(http.StatusFound, fmt.Sprintf("/healthauthority/%v", healthAuthority.ID))
+	}
+}
+
+type healthAuthorityFormData struct {
+	Issuer         string `form:"issuer"`
+	Audience       string `form:"audience"`
+	Name           string `form:"name"`
+	EnableStatsAPI bool   `form:"enable-stats-api"`
+	JwksURI        string `form:"jwks-uri"`
+}
+
+func (f *healthAuthorityFormData) PopulateHealthAuthority(ha *model.HealthAuthority) {
+	ha.Issuer = f.Issuer
+	ha.Audience = f.Audience
+	ha.Name = f.Name
+	ha.EnableStatsAPI = f.EnableStatsAPI
+	ha.SetJWKS(f.JwksURI)
+}
+
+type keyhealthAuthorityFormData struct {
+	Version  string `form:"version"`
+	PEMBlock string `form:"public-key-pem"`
+	FromDate string `form:"from-date"`
+	FromTime string `form:"from-time"`
+	ThruDate string `form:"Thru-date"`
```

Comment:
Did you intend to capitalize it to "Thru-date", or did you mean "thru-date"?

Reply:
This is most likely a bug, but again, this PR is just moving existing code around. To reduce the size of a diff and to reduce the chance of introducing a bug, I'm only moving code, no refactors or other changes in this PR.